ocaml/networkd/test/test_network_device_order.ml

```diff
@@ -0,0 +1,484 @@
+  Alcotest.(check bool) "present" true (present_of_mac mac_addr1 order) ;
+  Alcotest.(check bool) "present" true (present_of_mac mac_addr2 order) ;
+  Alcotest.(check bool) "present" true (present_of_mac mac_addr3 order) ;
+  Alcotest.(check bool) "present" true (present_of_mac mac_addr4 order) ;
+  Alcotest.(check bool) "present" true (present_of_mac mac_addr5 order)
+
+let test_initial_rules_via_label () =
+  let rules =
+    Rule.
+      [
+        {position= 0; index= Label name5}
+      ; {position= 1; index= Label name4}
+      ; {position= 2; index= Label name3}
+      ; {position= 3; index= Label name2}
+      ; {position= 4; index= Label name1}
+      ; {position= 5; index= Label name0}
+      ]
+  in
+  let order = sort' ~currents ~rules ~last_order:[] in
+  Alcotest.(check bool) "is Ok" true (Result.is_ok order) ;
+  let order = Result.get_ok order in
+  Alcotest.(check int) "6 devices in the order" 6 (List.length order) ;
+  Alcotest.(check int) "position assigned" 5 (pos_of_mac mac_addr0 order) ;
+  Alcotest.(check int) "position assigned" 4 (pos_of_mac mac_addr1 order) ;
+  Alcotest.(check int) "position assigned" 3 (pos_of_mac mac_addr2 order) ;
+  Alcotest.(check int) "position assigned" 2 (pos_of_mac mac_addr3 order) ;
+  Alcotest.(check int) "position assigned" 1 (pos_of_mac mac_addr4 order) ;
+  Alcotest.(check int) "position assigned" 0 (pos_of_mac mac_addr5 order) ;
+
+  Alcotest.(check bool) "present" true (present_of_mac mac_addr0 order) ;
+  Alcotest.(check bool) "present" true (present_of_mac mac_addr1 order) ;
+  Alcotest.(check bool) "present" true (present_of_mac mac_addr2 order) ;
+  Alcotest.(check bool) "present" true (present_of_mac mac_addr3 order) ;
+  Alcotest.(check bool) "present" true (present_of_mac mac_addr4 order) ;
+  Alcotest.(check bool) "present" true (present_of_mac mac_addr5 order)
+
+let test_replacement () =
+  let mac_addr0' = Macaddr.of_string "fc:f4:bb:e6:d7:b8" |> Result.get_ok in
+  let mac_addr1' = Macaddr.of_string "fc:f4:bb:e6:d7:b9" |> Result.get_ok in
+  let open Dev in
+  let dev0 =
+    {
+      name= "eno10"
+    ; pci= pci_addr0
+    ; mac= mac_addr0'
+    ; bios_eth_order= 1
+    ; multi_nic= false
+    }
+  in
+  let dev1 =
+    {
+      name= "eno11"
+    ; pci= pci_addr1
+    ; mac= mac_addr1'
+    ; bios_eth_order= 0
+    ; multi_nic= false
+    }
+  in
+  let currents =
+    currents
+    |> List.filter (fun dev -> dev.mac <> mac_addr0)
+    |> List.filter (fun dev -> dev.mac <> mac_addr1)
+    |> List.cons dev0
+    |> List.cons dev1
+  in
+  let order = sort' ~currents ~rules:[] ~last_order in
+  Alcotest.(check bool) "is Ok" true (Result.is_ok order) ;
+  let order = Result.get_ok order in
+
+  Alcotest.(check int) "6 devices in the order" 6 (List.length order) ;
+
+  Alcotest.(check int) "position assigned" 0 (pos_of_mac mac_addr0' order) ;
+  Alcotest.(check int) "position assigned" 1 (pos_of_mac mac_addr1' order) ;
```

Comment:
These positions don't match the bios_eth_order in currents, is this expected?

Reply:
Yes. As they are replacements in which case the `last order` should be used.